rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+    activation
+*   The whole model is trained offline
+*   The user want to retrain the 2 dense layers using personalized data on
+    device
+
+### Authoring TensorFlow graph
+
+**Defining the model using Keras Layers**
+
+The code defines the entire model using Keras API. Since we want to be retrain
+only dense layers on device, the model is defined in 2 parts (`conv_layers` and
+`dens_layer`).
+
+```python
+conv_layers = tf.keras.Sequential([
+  tf.keras.layers.Conv2D(10, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Conv2D(3, kernel_size=(3, 3), activation="relu"),
+  tf.keras.layers.Flatten(),
+])
+
+dense_layers = tf.keras.Sequential([
+  tf.keras.layers.Dense(50),
+  tf.keras.layers.Dense(10, activation='softmax'),
+])
+
+model = tf.keras.Sequential([conv_layers, dense_layers])
+```
+
+**Creating TensorFlow Functions for training and inference**
+
+With TensorFlow 2.0, the recommended way to use TensorFlow Lite is: Define a
+TensorFlow function for each invocable behavior in TensorFlow Lite. The
+converter will convert each TensorFlow function to a TensorFlow Lite subgraph,
+and users can choose subgraphs to invoke.
+
+The following code defines `inference` and `train` TensorFlow functions.
+
+```python
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32)])
+def inference(x):
+  return model(x)
+
+_LOSS_FN = tf.keras.losses.mean_squared_error
+_OPTIMIZER = tf.optimizers.RMSprop()
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32),
+    tf.TensorSpec(shape=[None, 10], dtype=tf.float32),
+])
+def train(x, y):
+  with tf.GradientTape() as tape:
+    prediction = model(x)
+    loss = _LOSS_FN(prediction, y)
+  gradients = tape.gradient(loss, model.trainable_variables)
+  _OPTIMIZER.apply_gradients(zip(gradients, model.trainable_variables))
+```
+
+The `train` function can be used to train the model offline. For retraining only
+dense layers, define another `train_dense_layers` TensorFlow function:
+
+```python
+
+@tf.function(input_signature=[
+    tf.TensorSpec(shape=[None, 64, 64, 3], dtype=tf.float32),
+    tf.TensorSpec(shape=[None, 10], dtype=tf.float32),
+])
+def train_dense_layers(x, y):
+  activation = conv_layers(x)
+  # Note: Gradient Tape is calculated only over the dense layers.
+  with tf.GradientTape() as tape:
+    prediction = dense_layers(activation)
+    loss = _LOSS_FN(prediction, y)
+  # Note: Gradients are only applied to trainable variables in dense layers.
+  gradients = tape.gradient(loss, dense_layers.trainable_variables)
+  _OPTIMIZER.apply_gradients(zip(gradients, dense_layers.trainable_variables))
+```
+
+Note that `tf.GradientTape` is calculated over only the dense layers, and
+`tape.gradients` are only applied to `dense_layers.trainable_variables` too.
+
+Though this example is simple, it's easy to extend it to support complex use
+cases. For example:
+
```

Comment:
Given that we're focusing on Keras models, it would be nice if there were a way to simply pass the model and use the built-in methods by default, rather than redefining loops. In other words, is it worth having a special happy path for Keras models?

Reply:
Thanks for the comment. 

Special happy Keras path sounds great. I'd love to continue this conversation. I don't fully understand "use the built-in methods by default". Do you have concrete recommendation for the design?

On the other hand, we're not only focusing on Keras model. The example is written with Keras API because it's more concise, but we're also aiming to support models constructed with lower-level TF APIs.

Reply:
Eg, model.fit/test/predict: https://www.tensorflow.org/beta/guide/keras/overview#train_from_numpy_data . I thought @gargn had added direct support to the converter as well, with from_keras_model or similar. An analogous API would be nice here-- given a keras model, assume we want to use .fit/.test/.predict for train/eval/inference.

Reply:
I see what you meant. I'm working closely with @gargn and I'm familiar with the TFLiteConverter code. I can prototype around this idea, but probably need to do that after the review period. 

One extra thought: We often want to train only a few layers (but not the entire model) on-device. I'm unsure if we can use the existing `Layer.trainable` flag to do this (training the entire model offline, turn off `trainable` for most layers, then convert the model.